console-subscriber/src/aggregator/mod.rs

```diff
@@ -689,63 +703,61 @@ impl Aggregator {
-                        // us to send only the new ones much like the IdData map.
-                        let mut async_op_stats = self.async_op_stats.update_or_default(async_op_id);
-                        async_op_stats.poll_stats.polls += 1;
-                        async_op_stats.latest_poll_op = Some(metadata.into());
-                        if async_op_stats.task_id.is_none() {
-                            async_op_stats.task_id = Some(task_id);
-                        }
-                        if async_op_stats.resource_id.is_none() {
-                            async_op_stats.resource_id = Some(resource_id.clone());
-                        }
+                let async_op_id = self.task_ids.id_for(async_op_id);
+                let resource_id = self.task_ids.id_for(resource_id);
+                let task_id = self.task_ids.id_for(task_id);
+
+                let mut async_op_stats = self.async_op_stats.update_or_default(async_op_id);
+                async_op_stats.poll_stats.polls += 1;
+                async_op_stats.task_id.get_or_insert(task_id);
+                async_op_stats.resource_id.get_or_insert(resource_id);
+
+                if matches!(readiness, Readiness::Pending)
+                    && async_op_stats.poll_stats.first_poll.is_none()
+                {
+                    async_op_stats.poll_stats.first_poll = Some(at);
+                }
 
-                        if let Readiness::Pending = readiness {
-                            if async_op_stats.poll_stats.first_poll.is_none() {
-                                async_op_stats.poll_stats.first_poll = Some(at);
-                            }
+                let poll_op = proto::resources::PollOp {
+                    metadata: Some(metadata.into()),
+                    resource_id: Some(resource_id.into()),
+                    name: op_name,
+                    task_id: Some(task_id.into()),
+                    async_op_id: Some(async_op_id.into()),
+                    readiness: match readiness {
+                        Readiness::Pending => proto::Readiness::Pending,
+                        Readiness::Ready => proto::Readiness::Ready,
+                    } as i32,
+                };
+
+                self.all_poll_ops.push(poll_op.clone());
+                self.new_poll_ops.push(poll_op);
+            }
+
+            Event::StateUpdate {
+                resource_id,
+                update,
+                ..
+            } => {
+                let resource_id = self.task_ids.id_for(resource_id);
+                if let Some(mut stats) = self.resource_stats.update(&resource_id) {
+                    let upd_key = (&update.val).into();
+                    match stats.attributes.get_mut(&upd_key) {
+                        Some(attr) => update_attribute(attr, update),
+                        None => {
+                            stats.attributes.insert(upd_key, update.into());
                         }
```

Comment:
style nit: i think it might be slightly nicer if we did this using the [`HashMap::entry`](https://doc.rust-lang.org/std/collections/hash_map/struct.HashMap.html#method.entry) API:
```suggestion
                    stats.attributes
                        .entry(upd_key)
                        .and_modify(|attr| update_attribute(attr, update))
                        .or_insert_with(|| update.into());
```

This has the advantage of only doing a single hashmap lookup, instead of two (one for `get_mut` and a second one for `insert` if the key is not present).

Reply:
Seems that requires cloning the `update` as it is moved into the `and_modify` closure.

Reply:
Okay, I believe we could alternatively still use a `match` with the `entry` API, like this:
```rust
                    match stats.attributes.entry(upd_key) {
                        Entry::Occupied(ref mut attr) => {
                            update_attribute(attr.get_mut(), update);
                        }
                        Entry::Vacant(attr) => {
                            attr.insert(update.into());
                        }
                    }
```
This still avoids performing a second lookup.

Obviously, not a blocker!